### PR TITLE
[FEATURE] Contextual help messages for node types and node properties

### PR DESCRIPTION
**Pull request by dimaip**
Allows to show contextual help messages in a popover element for node
types and node properties.
Supports markdown for styling message text.

Usage:
For node types:
```
'My.Package:SpecialHeadline':
  ui:
    help:
      message: 'Special headline displays text in **gold color**.'
```

For node properties:
```
'My.Package:SpecialHeadline':
  properties:
    title:
      ui:
        help:
          message: 'Lorem ipsum `dolor` sit amet'
```

Releases: master
Resolves: NEOS-1303

### CODE REVIEW

**Comment by dimaip** [comment]
:+1: from me and @ryzy (from gerrit).

File: TYPO3.Neos/composer.json [hunk 1]
```diff
@@ -10,7 +10,8 @@
         "typo3/setup": "*",
         "typo3/twitter-bootstrap": "*",
         "typo3/typo3cr": "*",
-        "typo3/typoscript": "*"
+        "typo3/typoscript": "*",
+        "league/commonmark": "*"
```
Comment (skurfuerst) [inline TYPO3.Neos/composer.json]: you also need to add league/commonmark to the merged composer.json

**Comment by skurfuerst** [comment]
Hey,

I did not test this change yet. On first glance, looks like a really nice feature :-)

2 Problems still come to my mind:

1. Translation: We should ensure the help messages are translatable as well, using the usual "i18n" string
2. Examples: Could you add some help texts for the standard node types here in this change as well? Makes it a lot easier to test I think :-)


All the best and keep up the great work,
Sebastian

**Comment by dimaip** [comment]
@skurfuerst thanks, will take care. Any suggestions for places where help messages would be most appropriate? :)

File: TYPO3.Neos/composer.json (hunk 1 above)
Comment (kdambekalns) [inline TYPO3.Neos/composer.json]: Which can easily be done with

```
cd Packages/Neos
php ../../Build/BuildEssentials/ComposerManifestMerger.php
```

**Comment by kdambekalns** [comment]
@dimaip I just worked in this a bit, moving some code around and implementing translation of help texts.

If you give me push access to your fork, I could just add to this PR, otherwise we'd need to coordinate in some other way. What do you think?

For now I pushed this to my fork, here is the diff to your version: https://github.com/dimaip/neos-development-collection/compare/help...kdambekalns:dimaip-help

**Comment by kdambekalns** [comment]
Ok, updated the PR with my changes. Enjoy the weekend! :)

**Comment by kdambekalns** [comment]
Note: the CI builds fail because the new dependency to league/commonmark is not picked up. So the result can (sadly) safely be ignored.

**Comment by dimaip** [comment]
@kdambekalns regarding the lose of hover on node type title: yes it's not so nice. I don't know how to easily fix it. What I'm sure about, is that the popup should definitely not be closed on the lose of hover, as the text message can be quite long, have scroll, clickable links inside etc.
I'd merge this change the way it is, if nobody comes up with a better idea how to fix this hover issue.

**Comment by kdambekalns** [comment]
Fine with me, @robertlemke will check this once more and merge it if it's done.

**Comment by robertlemke** [comment]
@dimaip thank you for this very useful feature!
It works all fine, so :+1: in terms of functionality and implementation.

However, I'm not 100% happy with the styling yet. I tried to tweak it a little, but I'm not very good at finding the right spots in CSS to get the layout right. I'll try to explain what I mean by adding some screenshots. Just keep in mind that these are really nitpicks, aiming for a high quality solution also at the user interface level. If all of you guys prefer to merge this without the style changes, so be it - but I'd be glad if someone who is a bit better at styling than me would quickly iron out some of the points I mentioned.

In general, I wonder if we could find a better overall style for popovers. What I have in mind is a white background and dark font and maybe even rounded corners in order to underline the fact that its a tooltip. You can see that in the following screenshot, where the background and borders of the tooltip don't play so nice with the underlying dialog:

screen shot 2015-12-08 at 09 33 50

Also, the question mark icon is a bit off, vertically, compared to the node type icon:

screen shot 2015-12-08 at 08 57 50

In the inspector, the help button looks a bit too small and also not vertically aligned with the text:

screen shot 2015-12-08 at 09 35 24

I don't know what kind of thumbnails people will use in practice, but if you use a regular image in portrait mode, you might miss that there's actually some more text below. Maybe the popover height should be a bit higher in that case?

screen shot 2015-12-08 at 09 41 42

Also, I think that there could be a bit more space between the text and the thumbnail:

screen shot 2015-12-08 at 09 53 12

And just one more idea for optimisation: is it possible to set a default target for inline links? Would be great if they would open in a new tab or window, because they might be useful for pointing to some further documentation.

**Comment by kdambekalns** [comment]
Ok, I'll try to take care of at least some of those things, timeboxed to an hour.

**Comment by kdambekalns** [comment]
Pushed a few changes:
- image has some space below in popover
- popover is positioned below
- help icon position fixed
- all links have _blank as target

The popover height and styling as such are unchanged, that could go into a different change.

**Comment by kdambekalns** [comment]
Ok, fixed some StyleCI issues, force pushed that update…

**Comment by robertlemke** [comment]
@kdambekalns perfect, thank you. I think it was worth the extra hours even though nobody likes to spend them of course ....

Looks great now, really love that new feature!

:+1: 

(and if we want to have a different style for popups in general, could be determined in a separate change)

**Comment by kitsunet** [comment]
This breaks functional tests and travis says so....

**Comment by robertlemke** [comment]
@kitsunet because the dependency is missing, as @kdambekalns wrote further above

**Comment by kitsunet** [comment]
Nope, there is a schema failure as well. The merge test and tests after this break as well.

**Comment by kdambekalns** [comment]
@kitsunet Would you point me to a Travis build? The one linked for the merge commit above shows other failures, AFAICS…

**Comment by kitsunet** [comment]
@kdambekalns See https://travis-ci.org/neos/neos-development-collection/jobs/95574079
Check the composer update, it installed commonmark, the error is about expected configuration schema.

**Comment by kdambekalns** [comment]
Thanks. Indeed, a rather broken test (because the failure has nothing to do with supertypes being correct), that needs to be fixed anyway. :)

**Comment by robertlemke** [comment]
@kitsunet okay, thanks for catching the failing test. I actually didn't investigate it further because I thought it would only fail due to the missing dependency.

**Comment by kitsunet** [comment]
\o/

**Comment by kitsunet** [comment]
We need to solve this dependency thing with travis to avoid it. It's really deceptive

**Comment by kdambekalns** [comment]
See https://github.com/neos/neos-development-collection/pull/250

**Comment by ryzy** [comment]
Just realised that feature has been merged - thanks guys, great job, great feature ;-)

**Pull request merged**